core/pkg/artifacts/multipart.go

```diff
@@ -0,0 +1,207 @@
+package artifacts
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"runtime"
+	"time"
+
+	"github.com/wandb/wandb/core/internal/gql"
+	"github.com/wandb/wandb/core/internal/hashencode"
+	"github.com/wandb/wandb/core/internal/observability"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	S3MinMultiUploadSize = 2 << 30   // 2 GiB, the threshold we've chosen to switch to multipart
+	S3MaxMultiUploadSize = 5 << 40   // 5 TiB, maximum possible object size
+	S3DefaultChunkSize   = 100 << 20 // 100 MiB
+	S3MaxParts           = 10000
+)
+
+// createMultiPartRequest checks if the file size is large enough to use multipart upload.
+// If so, it computes the hash of each parts required for getting multipart upload url from server.
+// Otherwise, it returns nil, which is a valid input for the graphql CreateArtifactFileSpecInput.
+func createMultiPartRequest(logger *observability.CoreLogger, path string) ([]gql.UploadPartsInput, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get file size for path %s: %w", path, err)
```

Comment:
`%v` is better than `%w` in these cases, see the guidelines [here](https://google.github.io/styleguide/go/best-practices.html#:~:text=It%20is%20best%20to%20avoid%20using%20%25w%20unless%20you%20also%20document%20(and%20have%20tests%20that%20validate)%20the%20underlying%20errors%20that%20you%20expose.%20If%20you%20do%20not%20expect%20your%20caller%20to%20call%20errors.Unwrap%2C%20errors.Is%20and%20so%20on%2C%20don%E2%80%99t%20bother%20with%20%25w.)